draco/synthesis/mockcatalog.py

```diff
@@ -795,19 +809,213 @@ def process(self, cat):
         # Generate standard normal z errors
         z_err = self.rng.normal(size=cat_z.shape[0])
         # Multiply by appropriate sigma
-        z_err *= (
-            self.sigma * np.ones_like(cat_z)
-            if self.sigma_type == "sigma_z"
-            else self.sigma * (1 + cat_z)
-        )
+        if self.use_catalog_z_errors:
+            if not np.any(cat_z_err):
+                self.log.error(
+                    "Warning: no existing z_error information in catalog, so no z errors will be added"
+                )
+            z_err *= cat_z_err
+        elif self.sigma_type == "sigma_z":
+            z_err *= self.sigma
+        else:  # self.sigma_type == "sigma_z_over_1plusz"
+            z_err *= self.sigma * (1 + cat_z)
+
+        # Add errors to catalog redshifts
+        cat_z += z_err
+        # If not using existing z_error information, add errors in quadrature
+        # with existing stored errors
+        if not self.use_catalog_z_errors:
+            cat_z_err[:] = (z_err ** 2 + cat_z_err ** 2) ** 0.5
+
+        return cat
+
+
+class AddEBOSSZErrorsToCatalog(task.SingleTask, random.RandomTask):
+    """Add eBOSS-type random redshift errors to redshifts in a catalog.
+
+    See the docstrings for _{qso,elg,lrg}_velocity_error() for descriptions
+    of each redshift error distribution.
+
+    Note also that the errors are added to the catalog in place, such that
+    if the original catalog is subsequently used in a pipeline, it will
+    have the errors included.
+
+    Attributes
+    ----------
+    tracer : {"ELG"|"LRG"|"QSO"}
+        Generate redshift errors corresponding to this eBOSS sample.
+    """
+
+    tracer = config.enum(["QSO", "ELG", "LRG"])
+
+    def process(self, cat):
+        """Generate random redshift errors and add to redshifts in catalog.
+
+        Parameters
+        ----------
+        cat : :class:`containers.SpectroscopicCatalog`
+            Input catalog.
+
+        Returns
+        ----------
+        cat_out : :class:`containers.SpectroscopicCatalog`
+            Catalog with redshift errors added.
+        """
+
+        # Get redshifts from catalog
+        cat_z = cat["redshift"]["z"][:]
+        cat_z_err = cat["redshift"]["z_error"][:]
+
+        # Generate redshift errors for the chosen tracer
+        z_err = self._generate_z_errors(cat_z, self.tracer)
 
         # Add errors to catalog redshifts
         cat_z += z_err
         # Add errors in quadrature with existing stored errors
-        cat_z_err = (z_err ** 2 + cat_z_err ** 2) ** 0.5
+        cat_z_err[:] = (z_err ** 2 + cat_z_err ** 2) ** 0.5
```

Comment:
It's not obvious this makes any sense. `z_err` is a realisation of a random variable, but you want to add in an estimate of the distribution variance instead. That probably requires another method to calculate that.

Reply:
Agreed - this somehow snuck through a previous review, and I left it alone as I wasn't sure what would be best here. Do we want to save the standard deviation of the underlying distribution even if the distribution isn't Gaussian? What is the intended use of the `z_error` field - simply a reflection of a (Gaussian) errorbar on `z`?

Reply:
The point of the `z_error` field was simply that it was a column available in the SDSS database and it sounded useful to have in our analysis, but I don't think we've ever actually used it, and so we've no real restriction of what it should be.

I think in this case, probably the best thing would be to estimate an effective sigma for each type (even for the non-Gaussian distributions), but I don't think we're really going to need it immediately so I'd be happy just not modifying it in this task and making an issue to do something smarter at a future date.

Reply:
Sure, sounds good to me

Reply:
See https://github.com/radiocosmology/draco/issues/151